dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+                if (!appInfo.processHandle)
+                {
+                    g_connectionManager.m_activityMap.erase(iter);
+
+                    errorMessage = "Failed to OpenProcess after the app is activated. The app may have crashed. Process Id: " + std::to_string(processId);
+                    RETURN_LAST_ERROR();
+                }
+
+                appInfo.activationStatus = AppActivationStatus::ActivationSucceeded;
+
+                // Failure to register wait handle tracking the new process lifetime is not fatal and should ot send failed ActivateAppResult back to client.
+                HRESULT hrRegisterWait{ LOG_IF_WIN32_BOOL_FALSE_MSG(RegisterWaitForSingleObject(
+                    &appInfo.processLifetimeTrackerHandle, appInfo.processHandle.get(), ProcessTerminationCallback, &appInfo, INFINITE, WT_EXECUTEONLYONCE),
+                    "Failed to register process lifetime tracking thread after the app is activated. Process Id: %u", processId) };
+                if (SUCCEEDED(hrRegisterWait))
+                {
+                    appActivatedAndTracked = true;
+                }
+            }
+            else
+            {
+                // The process Id is already in the m_processIdMap, which means the app activation does not create a new process.
+                // Apps can be single instance, which will use the same process to respond to multiple activations. 
+                // Remove the activity entry from the m_activityMap as we will not track duplicated activities backed by the same process.
+                g_connectionManager.m_activityMap.erase(iter);
+            }
+        }
+
+        SendActivateAppResult(pdu.activity_id(), S_OK, processId, isNewInstance);
+        return S_OK;
+    }
+    CATCH_RETURN()
+
+    HRESULT KozaniDvcServer::ProcessAppTerminationNotice(_In_ Dvc::ProtocolDataUnit& pdu) noexcept try
+    {
+        auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+        RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), g_connectionManager.m_closing);
+
+        auto iter{ g_connectionManager.m_activityMap.find(pdu.activity_id()) };
+        if (iter == g_connectionManager.m_activityMap.end())
+        {
+            RETURN_HR_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), "The activity Id %I64u does not exist in the activity map", pdu.activity_id());
+        }
+
+        auto& appInfo{ iter->second };
+        appInfo.activationStatus = AppActivationStatus::TerminationRequestedFromClient;
+
+        if (appInfo.pid == 0)
+        {
+            // The app activation has not finished yet. It will be terminated immediately upon activation since it is 
+            // marked with AppActivationStatus::TerminationRequestedFromClient. 
+            return S_OK;
+        }
+
+        FAIL_FAST_IF_MSG(!appInfo.processHandle, "Process handle should be valid when pid is not zero");
```

Comment:
FAIL_FAST -- see other comment

Reply:
appInfo.pid and appInfo.processHandle are NOT from outside inputs. It is purely internal logics. So FAIL_FAST is appropriate.

Reply:
Ahhh. So it's effectively an assert, "dev made mistake they could-and-should have known to avoid". OK